File: day02/day02.go
package day02

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Solution for star 1 first attempt
func SolveStar1Naive() int {
	// Read the input file
	file, err := os.Open("day02/input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return -1
	}
	defer file.Close()

	safe := 0
	// Process the input file line by line
	var scanner = bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		numbers := strings.Split(line, " ")

		if checkIfSafe(numbers) {
			safe++
		}

	}

	return safe
}

// Solution for star 2 first attempt
func SolveStar2Naive() int {
	// Read the input file
	file, err := os.Open("day02/simpleinput.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return -1
	}
	defer file.Close()

	safe := 0
	// Process the input file line by line
	var scanner = bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		numbers := strings.Split(line, " ")

		if checkIfSafeWithOneBadLevel(numbers) {
			safe++
		}

	}

	return safe
}

func checkIfSafe(nums []string) bool {
	previous, _ := strconv.Atoi(nums[0])
	isAccending := true
	for i := 1; i < len(nums); i++ {
		curNum, _ := strconv.Atoi(nums[i])

		difference := curNum - previous

		if i == 1 {
			// difference is positive it is decending
			isAccending = difference > 0
		}

		if isAccending && difference <= 0 {
			// if accending and next number is not
			return false
		} else if !isAccending && difference >= 0 {
			// if decending and next number is not
			return false
		} else if utils.Abs(difference) > 3 {
			// if the next difference is 2 big
			return false
		}

		previous = curNum
	}

	return true
}

// Give up and brute force... so bad will revisit idk
func checkIfSafeWithOneBadLevel(nums []string) bool {
	if checkIfSafe(nums) {
		return true
	}

	for index := range nums {
		newSlice := append(nums[:index], nums[(index+1):]...)
		fmt.Println(newSlice)
		if checkIfSafe(newSlice) {
			return true
		}
	}
	return false
}

File: day02/day02.py
import re

def is_safe(nums):
    inc = nums[1] > nums[0]
    if inc:
        for i in range(1, len(nums)):
            diff = nums[i] - nums[i-1]
            if not 1 <= diff <= 3:
                return False
        return True
    else:
        for i in range(1, len(nums)):
            diff = nums[i] - nums[i-1]
            if not -3 <= diff <= -1:
                return False
        return True


def is_really_safe(nums):
    if is_safe(nums):
        return True
    for i in range(len(nums)):
        if is_safe(nums[:i] + nums[i+1:]):
            return True
    return False

def SolveStar1():
    # Open the file in read mode
    with open('day02/input.txt', 'r') as file:
        # Read each line in the file
        total = 0
        for line in file:
            level = [int(x.strip()) for x in line.split(" ")]
            total += is_safe(level)

        return total


def SolveStar2():
    # Open the file in read mode
    with open('day02/input.txt', 'r') as file:
        total = 0
        # Read each line in the file
        for line in file:
            level = [int(x.strip()) for x in line.split(" ")]
            total += is_really_safe(level)

    return total